feat: Add support for Daml

stefanobaghino opened the pull request:
#### Description
- Add a `daml` module that reads the currently used SDK version for a project using the one indicated in the `daml.yaml` file or overrides it with `DAML_SDK_VERSION`
- Add Daml support to the `package` module (uses the `version` key in the `daml.yaml` file)

#### Motivation and Context
Closes #3903

#### How Has This Been Tested?
- [x] I have tested using **MacOS**
- [ ] I have tested using **Linux**
- [ ] I have tested using **Windows**

Apart from the relevant automated tests, I opened a sample Daml project and touched the `sdk-version` and `version` keys in the `daml.yaml` file, as well as the `DAML_SDK_VERSION` environment variable to see whether the changes were picked up correctly. Also went into a sub-directory to see whether the information disappeared as expected. I further made sure that the example in the documentation works, as well as checking that every option works as expected.

#### Checklist:
- [x] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly.

stefanobaghino commented:
A couple of notes on changes I made on the side:
- initially I couldn't make progress because of some incompatibility between `git2` and another library. I ran `cargo update` and that fixed it, but I'm not sure that's for the best (first time I write Rust code, so I'm really not familiar with the environment).
- Clippy raised a warning in the `c` module, I fixed it along the way. Not sure of why it was there, I guess it's a byproduct of my `cargo update` above.

davidkna commented on src/modules/daml.rs:
```diff
@@ -0,0 +1,120 @@
+    let mut module = context.new_module("daml");
+    let config: DamlConfig = DamlConfig::try_load(module.config);
+
+    let is_daml_project = context.try_begin_scan()?.set_files(&[DAML_YAML]).is_match();
```
Please use `detect_`.... options like most other modules.

stefanobaghino replied:
Addressed by https://github.com/starship/starship/pull/4004/commits/91989796cdeee210c08cf29484d53d99dcf3f791

davidkna commented on src/modules/c.rs:
```diff
@@ -65,12 +65,12 @@ pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
-                                Err(_e) => None,
-                            },
-                        )?,
+                        c_compiler_info.split_whitespace().find(|word| {
+                            match Version::parse(word) {
+                                Ok(_v) => true,
+                                Err(_e) => false,
+                            }
+                        })?,
```
#4002 has a fix for this.

stefanobaghino replied:
Should I wait for it to be merged and rebase on top of it?

stefanobaghino replied:
Changes reverted by https://github.com/starship/starship/pull/4004/commits/da426f34b864b1e17636da64083c57f3b33d874e. I'll wait for #4002 to be merged and rebase on top of it, thanks for letting me know.

davidkna commented on src/modules/daml.rs:
```diff
@@ -0,0 +1,120 @@
+fn read_sdk_version_from_daml_yaml(context: &Context) -> Option<String> {
+    let file_contents = context.read_file_from_pwd(DAML_YAML)?;
+    let daml_yaml = yaml_rust::YamlLoader::load_from_str(&file_contents).ok()?;
+    let sdk_version = daml_yaml.first()?[DAML_SDK_VERSION].as_str()?;
```
It looks like it's safe to do with this type, but please use `get` as `Index` may panic. Perhaps also add a test covering missing `DAML_SDK_VERSION`.

stefanobaghino replied:
I'm not a Rust expert, so I may need some help with this. I had a look at the other uses of this library in this codebase and it looks like the `Index` implementation always returns `Yaml` and that absence is dealt with the `.as_...` methods. Do I understand correctly? Let me write a test as you asked to validate my hypothesis.

stefanobaghino replied:
Test added in https://github.com/starship/starship/pull/4004/commits/8f6e77d998bbb793d4ee4a1081eabc1a69cae98c. IIUC this means that `Index` is safe to use here, right?

davidkna replied:
Just being cautious, as some implementation of `Index` will panic on undefined values.

stefanobaghino replied:
Makes a lot of sense. Can I consider this and https://github.com/starship/starship/pull/4004#discussion_r878697031 resolved?

davidkna replied:
sure.

davidkna commented on src/modules/package.rs:
```diff
@@ -247,6 +247,15 @@ fn get_shard_version(context: &Context, config: &PackageConfig) -> Option<String
+    let file_contents = context.read_file_from_pwd("daml.yaml")?;
+
+    let daml_yaml = yaml_rust::YamlLoader::load_from_str(&file_contents).ok()?;
+    let raw_version = daml_yaml.first()?["version"].as_str()?;
```
Please use `.get("version")`.

stefanobaghino replied:
I believe the same applies here as https://github.com/starship/starship/pull/4004#discussion_r878699757, but happy to receive further feedback of that not being the case.

davidkna reviewed (changes requested):
Please also revert the changes to `Cargo.lock`, I'd prefer not to accept a change to `Cargo.lock` from a new contributor or minimize the changes with `cargo update -p`. I think by default cargo ignores the lockfile, so I'm not sure if this was the cause of your problem.

stefanobaghino commented:
> Please also revert the changes to `Cargo.lock`, I'd prefer not to accept a change to `Cargo.lock` from a new contributor or minimize the changes with `cargo update -p`. I think by default cargo ignores the lockfile, so I'm not sure if this was the cause of your problem.

Not sure of what was happening initially, I reverted the changes on the `Cargo.lock` file and the issue seems to have just gone away. 🤷🏻 Thanks for the feedback!

stefanobaghino commented:
> Please also revert the changes to `Cargo.lock`, I'd prefer not to accept a change to `Cargo.lock` from a new contributor or minimize the changes with `cargo update -p`. I think by default cargo ignores the lockfile, so I'm not sure if this was the cause of your problem.

Addressed by https://github.com/starship/starship/pull/4004/commits/b153d3a7d6bc78ef1ba7df776699cd789a62f477

stefanobaghino commented:
#4002 has been merged. I rebased on top of the latest `master` and squashed my contributions into https://github.com/starship/starship/pull/4004/commits/466fec78824c1431ebcaf6f44f10d0731d25e5f3.

stefanobaghino commented:
![anticipation](https://media.giphy.com/media/Fo1cy8mqGDvbjpJBB7/giphy.gif)

stefanobaghino commented:
Let me know if there's something I should or can do to have this merged, happy to help. 🙂

davidkna commented:
Thanks for the contribution @stefanobaghino

stefanobaghino commented:
> Thanks for the contribution @stefanobaghino

Thank you for helping me and allowing me to contribute!
